training/run-training-task.js

```diff
@@ -0,0 +1,120 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Almond
+//
+// Copyright 2019 The Board of Trustees of the Leland Stanford Junior University
+//
+// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
+//
+// See COPYING for details
+"use strict";
+
+const Tp = require('thingpedia');
+const events = require('events');
+
+const trainingJobModel = require('../model/training_job');
+const modelsModel = require('../model/nlp_models');
+const db = require('../util/db');
+
+const Tasks = require('./tasks');
+
+const Config = require('../config');
+
+class Task extends events.EventEmitter {
+    constructor(jobId, jobDir) {
+        super();
+        this.jobId = jobId;
+        this.jobDir = jobDir;
+
+        this.killed = false;
+    }
+
+    get language() {
+        return this.jobInfo.language;
+    }
+
+    async load() {
+        await db.withTransaction(async (dbClient) => {
+            this.info = await trainingJobModel.get(dbClient, this.jobId);
+            this.config = JSON.parse(this.jobInfo.config);
+
+            this.modelInfo = null;
+            if (this.info.model_tag !== null) {
+                this.modelInfo = (await modelsModel.getByTag(dbClient, this.language, this.info.model_tag))[0];
+                if (!this.modelInfo) {
+                    // the model was deleted since the job was scheduled, or some other weirdness
+                    throw new Error('The model this job refers to no longer exists');
+                }
+            }
+
+            this.forDevices = await trainingJobModel.readForDevices(dbClient, this.jobId);
+        }, 'repeatable read', 'read only');
+    }
+
+    kill() {
+        this.killed = true;
+        this.emit('killed');
+    }
+
+    async setProgress(value) {
+        let auth = Config.TRAINING_ACCESS_TOKEN ? `Bearer ${Config.TRAINING_ACCESS_TOKEN}` : null;
+        return Tp.Helpers.Http.post(`${Config.TRAINING_URL}/jobs/${this.jobId}/progress`,
```

Comment:
as discussed you may want consider to updating db directly here. Then you won't need to worry about training daemon being down during training.

Reply:
One problem with that is that the progress value here is task progress, but the database contains job progress. Some component needs to be responsible for converting one to the other.

Reply:
It appears job.setProgress and job.setMetrics directly updates the db with the task progress and metrics. Maybe I missed something in the code?

Reply:
Well, at the moment yes, because the only task that reports progress is `train`, but in the future it should combine the progress from `prepare-training-set` and `train` into a single progress value.